va/validation-authority.go

```diff
@@ -98,25 +100,66 @@ func verifyValidationJWS(validation *jose.JsonWebSignature, accountKey *jose.Jso
 	return nil
 }
 
-// Validation methods
-
-// setChallengeErrorFromDNSError checks the error returned from Lookup...
+// problemDetailsFromDNSError checks the error returned from Lookup...
 // methods and tests if the error was an underlying net.OpError or an error
-// caused by resolver returning SERVFAIL or other invalid Rcodes and sets
-// the challenge.Error field accordingly.
-func setChallengeErrorFromDNSError(err error, challenge *core.Challenge) {
-	challenge.Error = &core.ProblemDetails{Type: core.ConnectionProblem}
+// caused by resolver returning SERVFAIL or other invalid Rcodes and returns
+// the relevant core.ProblemDetails.
+func problemDetailsFromDNSError(err error) *core.ProblemDetails {
+	problem := &core.ProblemDetails{Type: core.ConnectionProblem}
 	if netErr, ok := err.(*net.OpError); ok {
 		if netErr.Timeout() {
-			challenge.Error.Detail = "DNS query timed out"
+			problem.Detail = "DNS query timed out"
 		} else if netErr.Temporary() {
-			challenge.Error.Detail = "Temporary network connectivity error"
+			problem.Detail = "Temporary network connectivity error"
 		}
 	} else {
-		challenge.Error.Detail = "Server failure at resolver"
+		problem.Detail = "Server failure at resolver"
 	}
+	return problem
 }
 
+// getAddr will query for all A records associated with hostname and return
+// the first net.IP in the addrs slice and all addresses resolved.
+func (va ValidationAuthorityImpl) getAddr(hostname string) (addr net.IP, addrs []net.IP, problem *core.ProblemDetails) {
+	addrs, _, err := va.DNSResolver.LookupHost(hostname)
+	if err != nil {
+		problem = problemDetailsFromDNSError(err)
+		va.log.Debug(fmt.Sprintf("%s DNS failure: %s", hostname, err))
+		return
+	}
+	if len(addrs) == 0 {
+		problem = &core.ProblemDetails{
+			Type:   core.UnknownHostProblem,
+			Detail: fmt.Sprintf("Could not resolve %s", hostname),
+		}
+		return
+	}
+	addr = addrs[0]
+	va.log.Info(fmt.Sprintf("Resolved addresses for %s [using %s]: %s", hostname, addr, addrs))
+	return
+}
+
+// resolveAndConstructDialer gets the prefered address using va.getAddr and returns
+// the chosen address and dialer for that address and correct port.
+func (va ValidationAuthorityImpl) resolveAndConstructDialer(name string, scheme string) (func(string, string) (net.Conn, error), net.IP, []net.IP, *core.ProblemDetails) {
```

Comment:
A method signature this complicated is a sign something has gone wrong.  I think you've got the layering wrong here.  Rather than gluing these two things together, have the callers use `getAddr` to get the address, then `newDialer` to make the dialer.  As noted above, `getAddr` should be `getAddrs` and return the array; then `newDialer` should do the address selection and return the selected address and a dialer.  For the last thing, I would have a slight preference for wrapping those things in an object, e.g.:

```
type dialer struct {
  remoteIP net.IP
}
func (d *dialer) Dial(string, string) (net.Conn, error)
```